src/tree.jl

```diff
@@ -16,30 +16,30 @@ export Node,
 #     Any missing links (e.g., "c" does not have a sibling) link back to itself (c.sibling == c)
 # With this organization, no arrays need to be allocated.
 
-type Node{T}
+mutable struct Node{T}
     data::T
     parent::Node{T}
     child::Node{T}
     sibling::Node{T}
     
     # Constructor for the root of the tree
-    function (::Type{Node{T}}){T}(data::T)
+    function (::Type{Node{T}})(data::T) where T
         n = new{T}(data)
         n.parent = n
         n.child = n
         n.sibling = n
         n
     end
     # Constructor for all others
-    function (::Type{Node{T}}){T}(data::T, parent::Node)
+    function (::Type{Node{T}})(data::T, parent::Node) where T
         n = new{T}(data, parent)
         n.child = n
         n.sibling = n
         n
     end
 end
-Node{T}(data::T) = Node{T}(data)
-Node{T}(data::T, parent::Node{T}) = Node{T}(data, parent)
+Node(data::T) where {T} = Node{T}(data)
+Node(data::T, parent::Node{T}) where {T} = Node{T}(data, parent)
```

Comment:
Now that I know more I'd write this as `Node(data, parent::Node{T}) where T` (dropping the `::T` on `data`). Having the parent is enough to specify the `T` we need, and any `data` that can be converted to `T` will then work.

Completely optional, however.